### PR TITLE
ci(vllm): upgrade CI to vllm>=0.22.1 (cu13 stack, flashinfer jit-cache, Qwen3 embedding)

### PR DESCRIPTION
**Pull request by CatherineSue**
### Problem

CI was pinned to `vllm<0.19.1` because vLLM 0.19.1 bundled transformers v5, which broke `intfloat/e5-mistral-7b-instruct` embedding quality (last-token pooling loses the trailing EOS; self-similarity drops to ~0.33). That pin blocked every newer vLLM, and its companion `nixl<1.1.0` pin is incompatible with vLLM ≥ 0.22 (which requires `nixl>=1.1.0` for PD disaggregation).

The regression is unfixed upstream as of 0.22.1 — confirmed both statically (vLLM's own "numerical regression with transformers v5" pooling-test skips are still in place) and empirically on this PR (e5-mistral self-similarity 0.3341 under transformers 5.10.2).

### Solution

Floor instead of pin — `vllm>=0.22.1` — and fix each layer the upgrade surfaced, validated by CI runs:

1. **CUDA-matched torch stack**: install with `uv --torch-backend=auto`. Plain PyPI resolution pulls torch's cu13 default; `auto` matches the pod driver (the runner pool now supports CUDA 13, so the old "cu13 breaks ARC pods" premise behind the nixl pin is obsolete).
2. **NIXL install normalized like vLLM's own CI**: the bare `nixl` metapackage pulls both cu12 and cu13 backend wheels, so install the top-level shim with `--no-deps` plus the backend matching `torch.version.cuda`. The backend wheel also ships `nixl_ep` (MoE EP all-to-all, unused in CI) which vLLM imports eagerly whenever present — remove it post-install, and fail at install time via an import canary instead of mid-e2e.
3. **FlashInfer JIT cache**: vLLM 0.22.1 JIT-compiles FlashInfer kernels at engine startup (fused MoE for Qwen3-30B-A3B, attention ops for dense models) and the pods have no CUDA toolchain. Install `flashinfer-jit-cache` from flashinfer.ai matched to the flashinfer/CUDA versions — the same recipe as vLLM's Dockerfile.
4. **Embedding model swap**: e5-mistral cannot survive transformers v5, so `e2e-1gpu-embeddings` now uses `Qwen/Qwen3-Embedding-0.6B` — keeps last-token-pooling coverage, upstream-tested green under transformers v5 with an absolute MTEB score, documented SGLang embedding model (the spec is shared by both engine legs), and shrinks the model download from ~14GB to ~1.2GB.
5. **Honest relevance-score tolerance**: the relevance test compares against an fp32 CPU sentence-transformers reference; the old atol (5e-4 in cosine terms) was below cross-implementation noise — an fp32-serving experiment proved two whitespace-heavy reference docs diverge ~3e-3 cosine between vLLM and sentence-transformers regardless of dtype. The tolerance is now cosine 5e-3, still 2x tighter than upstream vLLM's own per-vector norm (1e-2) and 40x below the regression signature the test guards.

## Changes

- `scripts/ci_install_vllm.sh`: `vllm>=0.22.1` floor with `--torch-backend=auto`; nixl shim + CUDA-matched backend (`nixl-cu${CUDA_MAJOR}`) replacing `nixl<1.1.0`; `nixl_ep` removal; fatal import canary; `flashinfer-jit-cache` install; rewritten pin-rationale comments.
- `e2e_test/infra/model_specs.py`: embedding spec swapped to `Qwen/Qwen3-Embedding-0.6B` (tp 1).
- `e2e_test/embeddings/test_basic.py`, `test_correctness.py`: model markers, reference lookup, and prefix comments updated for the new model.

## Test Plan

- All vLLM serving e2e jobs green on vllm 0.22.1 + torch 2.11+cu130: `e2e-1gpu-chat` (incl. Qwen3-30B-A3B MoE via the JIT cache), `e2e-1gpu-completions`, `e2e-1gpu-gateway`, `e2e-1gpu-responses`, and `e2e-2gpu-pd` with real NIXL transfer on nixl-cu13.
- `e2e-1gpu-embeddings (sglang)` green with the new model; on the vllm leg, semantic-similarity passes (the transformers-v5 regression signature is gone) and the relevance subtest passes with the recalibrated tolerance.
- Known infra flakes observed during iteration, unrelated to this change: one 2-GPU node with broken IB state (UCX `rc_verbs` / Mooncake init failures) and one dirty-GPU abort; both cleared on rerun on healthy nodes.

<details>
<summary>Checklist</summary>

- [x] `cargo +nightly fmt` passes (no Rust changes)
- [x] `cargo clippy --all-targets --all-features -- -D warnings` passes (no Rust changes)
- [x] (Optional) Documentation updated (pin-rationale comments in the CI script)
- [ ] (Optional) Please join us on Slack [#sig-smg](https://slack.lightseek.org) to discuss, review, and merge PRs

</details>

### CODE REVIEW

File: scripts/ci_install_vllm.sh [hunk 1]
```diff
@@ -19,28 +19,20 @@ fi
...
-# runner pool catches up.
+# NIXL for vLLM PD disaggregation. vllm needs nixl >= 1.1.0, but the bare
+# metapackage installs both cu12 and cu13 wheels and the cu13 nixl_ep_cpp.so
+# breaks CUDA-12 pods — install the CUDA-matched wheel directly.
 echo "Installing nixl..."
-uv pip install "nixl<1.1.0"
+uv pip install "nixl-cu12>=1.2.0"
+
+# Import canary: fail here (not mid-e2e) if the nixl wheels mismatch the pod CUDA
+python3 -c "import nixl, nixl_ep" && echo "nixl import canary OK"
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply scripts/ci_install_vllm.sh]: Fixed in 505ea6b0 — split into separate statements so the import failure trips set -e. This mattered immediately: the first CI run hit a real nixl import failure that the && silently swallowed (see PR comment for the root cause).

[the remark this answers is not in the document]
Comment (CatherineSue) [reply scripts/ci_install_vllm.sh]: Fixed in 505ea6b0, same as the gemini thread — and the first CI run proved the point: a real import failure was swallowed by the && and resurfaced as worker crashes in three e2e jobs.

File: scripts/ci_install_vllm.sh [hunk 2]
```diff
@@ -19,28 +19,20 @@ fi
...
-# Install nixl for vLLM PD disaggregation (NIXL KV transfer).
-# Pin <1.1.0: 1.1.0 lands a cu13 nixl_ep_cpp.so that vLLM 0.19.0 imports
-# unconditionally, breaking CUDA-12 ARC pods. Drop once vLLM or the
-# runner pool catches up.
+# NIXL for vLLM PD disaggregation. vllm needs nixl >= 1.1.0, but the bare
+# metapackage installs both cu12 and cu13 wheels and the cu13 nixl_ep_cpp.so
+# breaks CUDA-12 pods — install the CUDA-matched wheel directly.
 echo "Installing nixl..."
-uv pip install "nixl<1.1.0"
+uv pip install "nixl-cu12>=1.2.0"
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply scripts/ci_install_vllm.sh]: Confirmed by wheel inspection and fixed in 505ea6b0 — nixl_cu12-1.2.0 ships only nixl_cu12 + nixl_ep, the top-level nixl shim lives in the meta wheel. Now installing the meta with --no-deps plus the cu12 backend, mirroring the end state of vLLM's own install-kv-connectors.sh. The first CI run also surfaced a second layer: the backend wheel's nixl_ep package is imported eagerly by vLLM when present and its .so needs a system CUDA runtime the pods lack, so it gets removed post-install (it's MoE EP all-to-all, unused by the KV connector).

File: scripts/ci_install_vllm.sh [hunk 3]
```diff
@@ -19,28 +19,27 @@ fi
...
-# force transformers back to 4.x. Not pinning transformers separately — if
-# we ship vllm we don't want to second-guess the library's own dep range;
-# drop this pin whenever vllm publishes a release that re-verifies embedding
-# correctness post-transformers-v5.
+# Floor 0.22.1: older vllm resolved an early transformers v5 that broke
+# e5-mistral last-token pooling (the old <0.19.1 pin); 0.22.1+ only admits
+# transformers >= 5.5.1. e2e-1gpu-embeddings is the quality gate.
 echo "Installing vLLM..."
-uv pip install "vllm<0.19.1"
+uv pip install "vllm>=0.22.1"
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply scripts/ci_install_vllm.sh]: Confirmed and fixed in 92596b70. The second CI run proved it: the install resolved torch==2.11.0 with the full cu13 stack (nvidia-cudnn-cu13, nccl-cu13, ...) and the canary failed with the nixl shim's own diagnostic — "torch reports CUDA 13 but nixl-cu13 is not installed". Now installing with uv --torch-backend=auto, vLLM 0.22.1's documented recipe, which matches the torch CUDA variant to the pod's driver (cu12x here); the vllm PyPI wheel itself is built against CUDA 12.9 so it pairs correctly.

File: scripts/ci_install_vllm.sh [hunk 4]
```diff
@@ -19,28 +19,27 @@ fi
...
+uv pip install "nixl-cu12>=1.2.0"
+
+# nixl-cu12 also ships nixl_ep (MoE all-to-all, unused here); vLLM imports it
+# eagerly when present and its .so needs a system CUDA runtime the pods lack
+rm -rf "$(python3 -c "import sysconfig; print(sysconfig.get_paths()['platlib'])")/nixl_ep"
+
+# Import canary: fail here (not mid-e2e) if the nixl install is broken
+# (torch first so its bundled CUDA libraries are loaded)
+python3 -c "import torch, nixl"
+echo "nixl import canary OK"
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply scripts/ci_install_vllm.sh]: Fixed in 5fd3c772 — the path is captured in a variable first, so a failed substitution aborts under set -e (assignment statements propagate the substitution's exit status) instead of being lost in the rm's argument.

File: scripts/ci_install_vllm.sh [hunk 5]
```diff
@@ -19,28 +19,29 @@ fi
...
-# we ship vllm we don't want to second-guess the library's own dep range;
-# drop this pin whenever vllm publishes a release that re-verifies embedding
-# correctness post-transformers-v5.
+# Floor 0.22.1: older vllm resolved an early transformers v5 that broke
+# e5-mistral last-token pooling (the old <0.19.1 pin); 0.22.1+ only admits
+# transformers >= 5.5.1. e2e-1gpu-embeddings is the quality gate.
+# --torch-backend=auto matches the torch CUDA variant to the pod's driver.
 echo "Installing vLLM..."
-uv pip install "vllm<0.19.1"
+uv pip install "vllm>=0.22.1" --torch-backend=auto
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply scripts/ci_install_vllm.sh:27]: The premise is stale: --torch-backend=auto resolved cu130 because the runner pool's driver supports CUDA 13 (it was upgraded since the original nixl<1.1.0 pin comment), so the cu13 packages are correct for these pods, not a mismatch. Empirically settled on this PR: with the coherent cu13 stack (torch 2.11.0+cu130, nixl-cu13, cudnn/cutlass-dsl cu13) every vLLM serving job is green — chat with the 30B MoE, completions, gateway, the embeddings serving path, and e2e-2gpu-pd with real NIXL transfer.

File: scripts/ci_install_vllm.sh [hunk 6]
```diff
@@ -19,28 +19,29 @@ fi
...
-# unconditionally, breaking CUDA-12 ARC pods. Drop once vLLM or the
-# runner pool catches up.
+# NIXL for vLLM PD disaggregation. The bare metapackage pulls both cu12 and
+# cu13 backends, so install the top-level shim alone, then the backend
+# matching torch's CUDA (same normalization as vLLM's own CI).
 echo "Installing nixl..."
-uv pip install "nixl<1.1.0"
+CUDA_MAJOR=$(python3 -c "import torch; print(torch.version.cuda.split('.')[0])")
+uv pip install --no-deps "nixl>=1.2.0"
+uv pip install "nixl-cu${CUDA_MAJOR}>=1.2.0"
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply scripts/ci_install_vllm.sh:35]: Declining: unlike the inline-substitution case in the other thread, a plain assignment VAR=$(cmd) takes the substitution's exit status, so under set -e a CPU-only torch (AttributeError on None.split) kills the script at exactly that line with a clear traceback — no masked failure and no malformed package name can reach uv. CPU-only torch is also unreachable here since the cu-flavored torch was installed one line earlier.

File: scripts/ci_install_vllm.sh [hunk 7]
```diff
@@ -19,28 +19,38 @@ fi
...
+echo "nixl import canary OK"
+
+# FlashInfer JIT cache: vLLM JIT-compiles flashinfer kernels at engine startup
+# and the pods have no CUDA toolchain — install the precompiled cache instead,
+# same recipe as vLLM's own Dockerfile.
+echo "Installing flashinfer-jit-cache..."
+CUDA_TAG=$(python3 -c "import torch; print(torch.version.cuda.replace('.', ''))")
+FLASHINFER_VERSION=$(python3 -c "import importlib.metadata as m; print(m.version('flashinfer-python'))")
+uv pip install "flashinfer-jit-cache==${FLASHINFER_VERSION}" \
+    --index-url "https://flashinfer.ai/whl/cu${CUDA_TAG}"
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply scripts/ci_install_vllm.sh:54]: Same rationale as the CUDA_MAJOR thread: assignments propagate command-substitution failures under set -e, so a missing flashinfer-python or CPU-only torch aborts the script at the exact line with python's traceback — clear enough for a CI script without extra validation scaffolding.
